packages/babel-plugin-bugfix-safari-class-field-initializer-scope/src/index.ts

```diff
@@ -0,0 +1,33 @@
+import { declare } from "@babel/helper-plugin-utils";
+import { types as t } from "@babel/core";
+import type { NodePath } from "@babel/core";
+
+function wrapInitializer(
+  path: NodePath<t.ClassProperty | t.ClassPrivateProperty>,
+) {
+  const { value } = path.node;
+
+  if (value && !(t.isLiteral(value) && !t.isTemplateLiteral(value))) {
```

Comment:
The condition for when wrapping is necessary is if there is anything wrapped in parentheses that is outside of a function, right?

To avoid too many IIFEs, I would expand this logic to, recursively:
- if NODE is a non-template literal, an Identifier, or a Function, return `false`
- if NODE is a CallExpression or a NewExpression, return `true` if recursing on the callee or on any of the arguments returns `true`
- if NODE is a TemplateLiteral, return `true` if recursing on the callee or on any of the inner expressions returns `true`
- if NODE is a TaggedTemplateExpression, return `true` if recursing on the tag or on the template returns `true`
- if NODE is an ArrayExpression, return `true` if recursing on any of the elements returns `true`
- if NODE is an ObjectExpression, return `true` if recursing on any of the property key/values returns `true`

This should avoid parens in most cases, except for when they are strictly necessary

Reply:
> The condition for when wrapping is necessary is if there is anything wrapped in parentheses that is outside of a function, right?

Not quite. The presence of parenthesis anywhere in the expression is enough to trigger the bug. The [check added in compat-table](https://github.com/compat-table/compat-table/commit/a40a7a2f93a4a07788a87a66115ac3e5138b9151) demonstrates this. `c = a[0]` is fine. `c = a[(0)]` is not 😬 

In [this comment](https://github.com/babel/babel/issues/14289#issuecomment-1047294289) of #14289, concern was raised about doing any transformation based on the presence (or not) of parenthesis. It's quite hard to do reliably.

That said, we can certainly cut down on the number of initializers which need to be transformed. Instead of gating it on parenthesis, we can look for any Identifiers which may refer to the parent scope. In fact, the logic we need is almost identical to what you wrote out, except that the last line should be:

- If NODE is an identifier, return <del>`false`</del><ins>`true`</ins> 

I've also added handling for a few other 'safe' nodes which can be skipped:

- added: "if NODE is a FunctionExpression or ArrowFunctionExpression, return `false`"
- added: "if NODE is a ThisExpression, return `false`"
- added: "if NODE is a SequenceExpression, return `true` if any of the elements returns `true`"